src/phy/sys/raw_socket.rs

```diff
@@ -1,81 +1,310 @@
-use std::{mem, io};
-use std::os::unix::io::{RawFd, AsRawFd};
-use libc;
-use super::*;
-
-#[derive(Debug)]
-pub struct RawSocketDesc {
-    lower: libc::c_int,
-    ifreq: ifreq
-}
 
-impl AsRawFd for RawSocketDesc {
-    fn as_raw_fd(&self) -> RawFd {
-        self.lower
+use phy::LinkLayer;
+use phy::sys;
+
+use std::io;
+use std::mem;
+use std::os::unix::io::RawFd;
+use std::os::unix::io::AsRawFd;
+
+cfg_if! {
+    if #[cfg(any(target_os = "macos", target_os = "freebsd"))] {
+        use std::ffi::CString;
+        use std::time::Duration;
     }
 }
 
-impl RawSocketDesc {
-    pub fn new(name: &str) -> io::Result<RawSocketDesc> {
-        let lower = unsafe {
-            let lower = libc::socket(libc::AF_PACKET, libc::SOCK_RAW | libc::SOCK_NONBLOCK,
-                                     imp::ETH_P_ALL.to_be() as i32);
-            if lower == -1 { return Err(io::Error::last_os_error()) }
-            lower
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RawSocket {
+    fd: sys::c_int,
+    dt: LinkLayer,
+    mtu: usize,
+    blen: usize,
+}
+
+#[cfg(all(target_os = "linux", target_env = "gnu"))]
+impl RawSocket {
+    pub fn with_ifname(ifname: &str) -> Result<RawSocket, io::Error> {
+        let flags = sys::if_name_to_flags(ifname).unwrap();
+        let link_layer = 
+            if flags & sys::IFF_LOOPBACK != 0 {
+                LinkLayer::Eth
+            } else if flags & sys::IFF_BROADCAST != 0 {
+                LinkLayer::Eth
+            } else if flags & sys::IFF_POINTOPOINT != 0 {
+                if flags & sys::IFF_NO_PI as i32 != 0 {
```

Comment:
i'm not sure this, when we create one `tun` interface with `IFF_NO_PI` flag on linux.